plugins/GSdx/GSOsdManager.cpp

```diff
@@ -0,0 +1,394 @@
+		pair.second.ty = (float)pair.second.bh / atlas_h;
+		pair.second.tw = (float)pair.second.bw / atlas_w;
+
+		x += pair.second.bw;
+	}
+
+	m_texture_dirty = false;
+}
+
+/* This is dumb in that it doesn't check for malformed UTF8. This function
+ * is not expected to operate on user input, but only on compiled in strings */
+void dumb_utf8_to_utf32(const char *utf8, char32_t *utf32, unsigned size) {
+	while(*utf8 && --size) {
+		if((*utf8 & 0xF1) == 0xF0) {
+			*utf32++ = (utf8[0] & 0x07) << 18 | (utf8[1] & 0x3F) << 12 | (utf8[2] & 0x3F) << 6 | utf8[3] & 0x3F;
+			utf8 += 4;
+		} else if((*utf8 & 0xF0) == 0xE0) {
+			*utf32++ =                          (utf8[0] & 0x0F) << 12 | (utf8[1] & 0x3F) << 6 | utf8[2] & 0x3F;
+			utf8 += 3;
+		} else if((*utf8 & 0xE0) == 0xC0) {
+			*utf32++ =                                                   (utf8[0] & 0x1F) << 6 | utf8[1] & 0x3F;
+			utf8 += 2;
+		} else if((*utf8 & 0x80) == 0x00) {
+			*utf32++ =                                                                           utf8[0] & 0x7F;
+			utf8 += 1;
+		}
+	}
+
+	if(size) *utf32 = *utf8; // Copy NUL char
+}
+
+void GSOsdManager::AddGlyph(char32_t codepoint) {
+	if(c_info.count(codepoint) == 0) {
+		m_texture_dirty = true;
+		c_info[codepoint]; // add it
+		if(FT_HAS_KERNING(m_face)) {
+			FT_UInt new_glyph = FT_Get_Char_Index(m_face, codepoint);
+			for(auto pair : c_info) {
+				FT_Vector delta;
+
+				FT_UInt glyph_index = FT_Get_Char_Index(m_face, pair.first);
+				FT_Get_Kerning(m_face, glyph_index, new_glyph, FT_KERNING_DEFAULT, &delta);
+				k_info[std::make_pair(pair.first, codepoint)] = delta.x >> 6;
+			}
+		}
+	}
+}
+
+void GSOsdManager::Log(const char *utf8, uint32 color) {
+	if(!m_log_enabled)
+		return;
+
+	char32_t buffer[256];
+	dumb_utf8_to_utf32(utf8, buffer, countof(buffer));
+	m_log.push_back(log_info{color, buffer, std::chrono::system_clock::time_point()});
+	for(char32_t* c = buffer; *c; ++c) AddGlyph(*c);
+}
+
+void GSOsdManager::Monitor(const char *key, const char *value, uint32 color) {
+	if(!m_monitor_enabled)
+		return;
+
+	if(value && *value) {
+		char32_t buffer[256], vbuffer[256];
+		dumb_utf8_to_utf32(key, buffer, countof(buffer));
+		dumb_utf8_to_utf32(value, vbuffer, countof(vbuffer));
+		m_monitor[buffer] = std::make_pair(vbuffer, color);
+		for(char32_t* c = buffer; *c; ++c) AddGlyph(*c);
+		for(char32_t* c = vbuffer; *c; ++c) AddGlyph(*c);
+	} else {
+		char32_t buffer[256];
+		dumb_utf8_to_utf32(key, buffer, countof(buffer));
+		m_monitor.erase(buffer);
+	}
+}
+
+void GSOsdManager::Indicate(std::string key, bool on) {
+	if(!m_indicator_enabled)
+		return;
+
+	m_indicator[key].on = on;
+}
+
+void GSOsdManager::RenderGlyph(GSVertexPT1* dst, glyph_info g, float x, float y, uint32 color) {
+	float x2 = x + g.bl * (2.0f/m_real_size.x);
+	float y2 = -y - g.bt * (2.0f/m_real_size.y);
+	float w = g.bw * (2.0f/m_real_size.x);
+	float h = g.bh * (2.0f/m_real_size.y);
+
+	dst->p = GSVector4(x2    , -y2    , 0.0f, 0.0f);
```

Comment:
factored how so?  This function defines two triangles.  two points from both triangles match to make a square, but I'm not sure how to make this any simpler than it is.